File: sorting-algos/radix_sort/radix_sort.py
def count_sort(arr,exp,size):   #radix sort is only applicable for non-negative integers
    aux = [0]*size          #contains sorted array
    freq=[0]*10             #frequency of digits 0 to 9
    for i in arr:
        temp = int(i/exp)   #exp is power of 10 according to digit used for sorting
        temp=temp%10        #temp stores the required digit of the number
        freq[temp]+=1
    for i in range(1,10):
        freq[i]+=freq[i-1]  #finding cumulative frequency for determining place to insert
    j=size-1
    while(j>=0):
        temp=int(arr[j]/exp)
        temp=temp%10
        aux[freq[temp]-1]=arr[j]    #inserting the element and decreasing frequency
        freq[temp]=freq[temp]-1
        j=j-1
    for i in range(0,size):
        arr[i] = aux[i]         #copying sorted array into original array


def radix_sort(arr,size):
    temp=max(arr)
    digit=0
    while(temp!=0):
        digit=digit+1
        temp=int(temp/10)
    exp=1
    for i in range(digit):
        count_sort(arr,exp,size)
        exp*=10

if(__name__=='__main__'):
    size=int(input('Enter the number of elements in array\n'))
    print('Enter the elements')
    arr=[int(input()) for i in range(size)]
    radix_sort(arr,size)
    print('The sorted array is')
    print(arr)

File: sorting-algos/radix_sort/radix_sort.cpp
#include <iostream>
using namespace std;

void counting_sort(int arr[],int n,int exp)  //radix sort is only applicable for non-negative integers
{
    int aux[n];      //contains sorted array
    int freq[10];       //frequency of digits 0 to 9
    int i;
    for(i=0;i<10;i++)
        freq[i]=0;
    int temp;
    for(i=0;i<n;i++)
    {
        temp=(arr[i]/exp)%10;   //exp is power of 10 according to digit used for sorting
        freq[temp]++;
    }

    for(i=1;i<10;i++)
        freq[i]+=freq[i-1];     //finding cumulative frequency for determining place to insert

    for(i=n-1;i>=0;i--)
    {
        temp=(arr[i]/exp)%10;
        aux[freq[temp]-1]=arr[i];   //inserting the element and decreasing frequency
        freq[temp]--;
    }

    for(i=0;i<n;i++)
        arr[i]=aux[i];              //copying sorted array into original array

}

void radix_sort(int arr[],int n)
{
    int i;
    int exp=1;
    int maxm=arr[0];
    for(i=1;i<n;i++)
    {
        if(maxm<arr[i])
            maxm=arr[i];
    }
    int digit =0;
    int temp=maxm;
    while(temp!=0)
    {
        digit++;
        temp/=10;
    }
    for(i=0;i<digit;i++)
    {
        counting_sort(arr,n,exp);
        exp*=10;
    }
}
int main ()
{
    int n,i;
    cout<<"Enter the number of elements in array"<<endl;
    cin>>n;
    int arr[n];
    cout<<"Enter the elements of array"<<endl;
    for(i=0;i<n;i++)
        cin>>arr[i];
    radix_sort(arr,n);
    cout<<"The sorted array is"<<endl;
    for(i=0;i<n;i++)
        cout<<arr[i]<<" ";
    cout<<endl;
}
